### ISSUE TITLE
Improve repo_config

### ISSUE THREAD

**Issue by gpsaggese** [causify-ai/helpers#88]

**Comment by heanhsok** [comment]
hi @gpsaggese 

Could I take this one? Is there existing work from before, or should I start from scratch?

Ideally, we want to use just one config file to store variables for both our Python and bash scripts. There are a few options such as `.yml`, `.ini`, `.toml`, and `.env`.

Given that we need to load variables in this file from both our Python and bash scripts, the easiest one we can use without having to write extra parsing logic (especially for our shell script) would be `.env`. The downside is that this format doesn't support complex datatypes and hierarchical data structure.

For example, 
- in the config file (`repo_config.env`)
```
IS_SUPER_REPO=1
```
- in our python file,
```
import dotenv
config = dotenv.dotenv_values("repo_config.env")
config['IS_SUPER_REPO']
```
- in our shell file,
```
source repo_config.env
echo $IS_SUPER_REPO 
```

WDYT?

**Comment by gpsaggese** [comment]
I started working on this and some early (very messy code) is https://github.com/causify-ai/helpers/pull/288

I suggest to
- read the code
- make a proposal of how the system should work
- discuss
- implement

**Comment by heanhsok** [comment]
How the system should work:

I'm mainly just describing the appraoch from the PR.

1) We want to store all configurations in a central location (which can be read by both the Python code and shell scripts).  
2) Given that .env files don't support nested structures and complex data types, a .yml file is a better option. Current configurations:  
   1) Repository identifier (name, GitHub account, host, etc.)  
   2) Docker (image name, etc.)  
   3) Runnable repository/directory configuration (use_helpers_as_submodule, dir_suffix, dir_tag etc.)  
   4) virtual environment for thin client  
   5) Shared data directory  
   6) S3 buckets  
   7) basic AWS config (AWS profile, account ID, ECR base path)  
3) The setenv.sh script reads the configuration file to bootstrap the thin environment:  
   1) We can use an additional library such as `jq` to parse the file  
   2) Or add our own parsing logic to the `thin_client_util`  
4) For the Python scripts, we can maintain the same interface as our current approach. However, instead of hard-coding variables in the code, we can read them from the `.yml` file.  
   1) Load the file as a Python dictionary (for example, using `yaml.safe_load`)  
   2) Use a helper class RepoConfig to manages retrieval from the `.yml` and provides access to these configuration variables

Implementation:

1) Implement according to the above spec  in //helpers
2) Add unit tests for the implementation
3) Test the changes from other repos (other than //helpers)
4) Test the changes from both local dev and CI pipelines
4) Once the changes are merged from //helpers, we can change to this new approach for all repos that use `//helpers`  
5) Apply the symbolic approach for more common files as we have now moved the part that is unique to each repo out of the code

**Comment by gpsaggese** [comment]
Specs are perfect.

We should have comments describing each var in the code. Yaml should support them natively.
Not sure if we need to organize vars with hierarchy, at least right now.

**Comment by gpsaggese** [comment]
- [ ] Code is complete, now running to make sure everything works 
- [ ] Add a note to the doc and to the paper @sonniki

**Comment by heanhsok** [comment]
### Testing Done
Thus far I have tested the changes these two repos. They passed the manual tests (e.g. running the invoke docker_bash, run_fast_tests, run_slow_tests) and tests on our CI. 
- //helpers: https://github.com/causify-ai/helpers/pull/323
- //cmamp: https://github.com/causify-ai/cmamp/pull/11482


### Summary of the changes
I have picked up only some necessary changes from https://github.com/causify-ai/helpers/pull/288 for repo_config improvement for further modification as there are some other changes related to DinD and Sibling containers (which can be addressed in separate PRs IMO)
 
This is sort of another big refactoring and this changes will need to be propagated to all the repos that uses helpers.
To summarize the main changes for our Python here are
- the `repo_config.py` is removed in favor of `repo_config.yml`
- the `henv.execute_repo_config_code("func")` layer here is also completely removed in favor of direct access to the config either through the interface from `hserver` (for server related config) or `repo_config_utils` (for repo related config)
- some configs variable are also moved between repo_config and hserver for better organization and reflection of their purposes

The changes are not yet mergable as we need to test it more first and plan for release once we are confident that they serve the required use cases as well as won't break anything important :)

Please let me know if you have any initial feedback on the above PRs @gpsaggese 

### Next actions:
- Test the changes on the //orange and runnable directories (which have a different setup than //cmamp and //helpers)
- Review existing unit tests and add additional test coverage (if necessary) related to these changes
- Update the bash script to use the configuration from the new setup

**Comment by gpsaggese** [comment]
Phenomenal work. 

- [ ] Let's get everything tested and merged ASAP, then we'll keep doing some clean up of this part.
- [ ] Let's document in the internal doc and in the paper the design. It doesn't have to be an essay, but just something that explains the logic of what we are doing (e.g., each repo has its own config, all the config should be there, etc). The code is so complicated (we passed the 1M LOC and then I stopped counting) that it's difficult to understand *WHY* something was done.

**Comment by heanhsok** [comment]
1) Testing on different setups is done in the following PRs on local dev and CI pipelines
  - //helpers: https://github.com/causify-ai/helpers/pull/323
  - //cmamp and its runnable dir https://github.com/causify-ai/cmamp/pull/11482
  - //orange: https://github.com/causify-ai/orange/pull/692
2) the `setenv.sh` is also updated to read from the config file

**Comment by gpsaggese** [comment]
Very well done. Let's merge all these guys. Can you do it? Should I do it? Is there an order? (I believe so)

**Comment by heanhsok** [comment]
Yes there's an order (helpers -> cmamp -> orange). I'll resolve the new comments and then we can merge them.

**Comment by heanhsok** [comment]
Closing this for now as the main changes are done and remaing works can be addressed in follow ups
- Documentation: https://github.com/causify-ai/helpers/issues/361
- Roll out changes to remaing repos (https://github.com/causify-ai/cmamp/issues/11561)
  - //lemonade: https://github.com/causify-ai/cmamp/issues/11579
  - //quant_dashboard: https://github.com/causify-ai/cmamp/issues/11542
